aten/src/TH/generic/THTensorRandom.cpp

```diff
@@ -215,7 +215,7 @@ void THTensor_(multinomialAliasSetup)(THTensor *probs, THLongTensor *J, THTensor
     {
       // sometimes an large index isn't added to J.
       // fix it by making the probability 1 so that J isn't indexed.
-      if(J_data[i] <= 0)
+      if(J_data[i] < 0)
```

Comment:
I think for safety, we can initialize `J` to -1 instead of 0. Then if J[i] continues to be -1 at this step, we can safely assign `q[i] = 1.0`. Otherwise it might be possible due to numerical issues that any additional probability mass gets default assigned to bucket 0. Even if that is not a serious issue, this will help with debugging, since we'll know immediately that anything with -1 was an excess bucket with no alias and should have q=1. and not 0.999, for instance.

I also think we should add more test cases to check that the alias table is correctly constructed, since these tests are cheap (no stochasticity), and will point to any other issues in the alias table construction that we might have missed out. e.g.

```python
    def test_alias_table(self):
        # Add more test cases
        for probs in [torch.tensor([0.2501, 0.25, 0.2499, 0.25]),
                      torch.tensor([0.8, 1.999, 0.001])]:
            alias_table, prob_table = torch._multinomial_alias_setup(probs)
            actual = torch.zeros(probs.shape)
            for i, vals in enumerate(zip(alias_table, prob_table)):
                idx, p = vals
                actual[i] += p
                actual[idx] += 1. - p
            actual = actual / len(probs)
            self.assertEqual(actual, probs, 1e-6)
```

Reply:
Let me know if you need any help, I'm happy to push this directly to your branch and debug any test failures.

Reply:
Thank you, I’ll let you know if I need help. I’ll make these changes and push.

Reply:
@vishwakftw - The alias table seems to be correctly constructed. There was a typo in my test case (corrected above) `1.99 --> 0.199`; was that the cause of the failure?

Reply:
No, I corrected the typo. This case passed, the first one didn’t.

Reply:
I see. I made an additional change that I mentioned above of initializing J to -1 `THLongTensor_fastSet1d(J, i, -1L);`, but I doubt if that should affect the results.